cleanrl/ppo_memory_env_lstm.py

```diff
@@ -0,0 +1,407 @@
+                    writer.add_scalar("charts/episodic_length", item["episode"]["l"], global_step)
+                    break
+
+        # bootstrap value if not done
+        with torch.no_grad():
+            next_value = agent.get_value(next_obs).reshape(1, -1)
+            if args.gae:
+                advantages = torch.zeros_like(rewards).to(device)
+                lastgaelam = 0
+                for t in reversed(range(args.num_steps)):
+                    if t == args.num_steps - 1:
+                        nextnonterminal = 1.0 - next_done
+                        nextvalues = next_value
+                    else:
+                        nextnonterminal = 1.0 - dones[t + 1]
+                        nextvalues = values[t + 1]
+                    delta = rewards[t] + args.gamma * nextvalues * nextnonterminal - values[t]
+                    advantages[t] = lastgaelam = delta + args.gamma * args.gae_lambda * nextnonterminal * lastgaelam
+                returns = advantages + values
+            else:
+                returns = torch.zeros_like(rewards).to(device)
+                for t in reversed(range(args.num_steps)):
+                    if t == args.num_steps - 1:
+                        nextnonterminal = 1.0 - next_done
+                        next_return = next_value
+                    else:
+                        nextnonterminal = 1.0 - dones[t + 1]
+                        next_return = returns[t + 1]
+                    returns[t] = rewards[t] + args.gamma * nextnonterminal * next_return
+                advantages = returns - values
+
+        # flatten the batch
+        b_obs = obs.reshape((-1,) + envs.single_observation_space.shape)
+        b_logprobs = logprobs.reshape(-1)
+        b_actions = actions.reshape((-1,) + envs.single_action_space.shape)
+        b_dones = dones.reshape(-1)
+        b_advantages = advantages.reshape(-1)
+        b_returns = returns.reshape(-1)
+        b_values = values.reshape(-1)
+
+        b_lstm_states = (
+            lstm_states[0].permute((1,0,2,3)).reshape(1, -1, agent.lstm.hidden_size),
+            lstm_states[1].permute((1,0,2,3)).reshape(1, -1, agent.lstm.hidden_size)
+        )
+
+        # Optimizaing the policy and value network
+        assert args.num_envs % args.num_minibatches == 0
+        envsperbatch = args.num_envs // args.num_minibatches
+        envinds = np.arange(args.num_envs)
+        flatinds = np.arange(args.batch_size).reshape(args.num_envs, args.num_steps)
+        clipfracs = []
+        for epoch in range(args.update_epochs):
+            # np.random.shuffle(b_inds)
+            np.random.shuffle(envinds)
+            for start in range(0, args.num_envs, envsperbatch):
+                end = start + envsperbatch
+                mbenvinds = envinds[start:end]
+                mb_inds = flatinds[mbenvinds].ravel()
+
+                _, newlogprob, entropy, newvalue, _ = agent.get_action_and_value(
```

Comment:
input to the lstm during update phase should be (seq, batch , feature) but it's  ( batch , seq, feature)

Reply:
Nice catch.

Reply:
Sorry I didn't quite get this. I have

```
b_lstm_states[0][:,mb_inds].shape
Out[2]: torch.Size([1, 256, 512])
```

which I assume is `(seq, batch, feature)`?

Reply:
Out[2]: torch.Size([1, 256, 512]) , 1 is not a sequence but a batch and 256 is a sequence

Reply:
I'm only refering to the update phase, inference is ok

Reply:
So during the inference, it has
```
next_lstm_state[0].shape
Out[3]: torch.Size([1, 8, 128])
```
Isn't 8 the batch in this case?

Reply:
yes, during inference you passs single sequence with a batch of 8, during update you concatenate two trajectories of length 128 to form single sequence of length 256 and batch of 1

Reply:
for the update you need to switch batch and seq_length

Reply:
I think I see wha you are saying, but shouldn’t it to have shape `(128, 2, 128)`? So basically the batch size is 2 and the sequence length is 128.

Reply:
sure